src/ucp/tag/rndv.c

```diff
@@ -622,6 +630,8 @@ static void ucp_rndv_send_frag_rtr(ucp_worker_h worker, ucp_request_t *rndv_req,
         freq->recv.state.dt.contig.md_map = 0;
         freq->recv.frag.rreq              = rreq;
         freq->recv.frag.offset            = offset;
+        ucp_rndv_set_recv_tag_info(freq, rreq->recv.tag.info.sender_tag,
```

Comment:
why is this needed for fragments? it worked without it before..
AFAIR pipelined RDNV is not used for TCP.. ?
seems we need to adjust frag size by sender length, in line 605?

Reply:
> seems we need to adjust frag size by sender length, in line 605?

looks like smth similar is done by #4450 (it's adjusted on a sender side)

> AFAIR pipelined RDNV is not used for TCP.. ?

this PR fixes it to not use pipelined RNDV for TCP when send or recv buffer isn't CUDA

> why is this needed for fragments? it worked without it before..

this is needed to have this value initialized and use it on `src/ucp/tag/rndv.c:99` as it common code for PIPELINE and PUT Zcopy

Reply:
> this PR fixes it to not use pipelined RNDV for TCP when send or recv buffer isn't CUDA

Can we skip this function altogether in this case?
it's weird that we call pipelined flow on receiver but make sender think it's regular rndv put

Reply:
> Can we skip this function altogether in this case?

then we need some indication in `ucp_tag_rndv_rtr_pack()` what the protocol will be used (RNDV pipeline or RNDV PUT Zcopy) to do smth like:
```
rndv_rtr_hdr->size    = (!rndv_pipeline ? ucs_min(rreq->recv.length,
                                                  rreq->recv.tag.info.length) : 
                         rreq->recv.length);
```

> it's weird that we call pipelined flow on receiver but make sender think it's regular rndv put

the receiver asks the sender to do RNDV pipeline scheme if receive buffer is CUDA and (`RNDV_SCHEME == auto`), RTR size in this case < the send buffer size in this case, so sender uses RNDV pipeline scheme.

Reply:
can we add length field to ucp_request::send.rndv_rtr ?

Reply:
> can we add length field to ucp_request::send.rndv_rtr ?

sure, this is under `union` where there are other structures that are larger (e.g. `struct rndv_put`)
done